### PR TITLE
 Fix auto_explain to handle Greenplum EXPLAIN handling

### PR DESCRIPTION
**Pull request by danielgustafsson**
When running EXPLAIN ANALYZE queries, the showstat context was being populated twice which led to an assertion failure. Fix by adding a flag to the struct to signal that the stats gathering has already
been performed. Also add the Greenplum Executor end statistics to the printing and properly set CDB instrumentation.

This is a re-working of #6085 done through the review process.

### CODE REVIEW

**Comment by leskin-in** [comment]
The description of https://github.com/greenplum-db/gpdb/pull/6100/commits/545a76f1e8df43c05757df06a56ec241ba518a56 is not correct. That commit closes https://github.com/greenplum-db/gpdb/issues/5906.

File: contrib/auto_explain/auto_explain.c [hunk 1]
```diff
@@ -286,6 +299,10 @@ explain_ExecutorEnd(QueryDesc *queryDesc)
 	{
 		double		msec;
 
+		/* Wait for completion of all qExec processes. */
+		if (queryDesc->estate->dispatcherState && queryDesc->estate->dispatcherState->primaryResults)
+			cdbdisp_checkDispatchResult(queryDesc->estate->dispatcherState, DISPATCH_WAIT_NONE);
```
Comment (hlinnaka) [inline contrib/auto_explain/auto_explain.c]: It's a bit sad that this is needed. I feel that we have a bad abstraction somewhere, shouldn't need to peek so deep into the internals in an extension. (But probably not worth a massive refactoring, if there's no easy fix...)

File: src/backend/commands/explain_gp.c [hunk 2]
```diff
@@ -1877,6 +1881,21 @@ cdbexplain_showExecStats(struct PlanState *planstate, ExplainState *es)
 	}
 }								/* cdbexplain_showExecStats */
 
+/*
+ *	ExplainPrintExecStatsEnd
+ *			External API wrapper for cdbexplain_showExecStatsEnd
+ *
+ * This is an externally exposed wrapper for cdbexplain_showExecStatsEnd such
+ * that extensions, such as auto_explain, can leverage the Greenplum specific
+ * parts of the EXPLAIN machiner.
```
Comment (hlinnaka) [inline src/backend/commands/explain_gp.c]: s/machiner/machinery/

File: contrib/auto_explain/auto_explain.c (hunk 1 above)
Comment (danielgustafsson) [reply contrib/auto_explain/auto_explain.c]: Agreed. If there is a better way it has eluded me, I think this is the only way for now.

File: src/backend/commands/explain_gp.c (hunk 2 above)
Comment (danielgustafsson) [reply src/backend/commands/explain_gp.c]: Thanks, will fix!

**Comment by danielgustafsson** [comment]
Pushed to master, thanks for review and thanks for your contribution @leskin-in !

**Pull request closed without merging**